test/array.js

```diff
@@ -795,4 +795,193 @@ describe('array', function () {
+                expect(value).to.deep.equal(['s1', 2]);
+                done();
+            });
+        });
+
+        it('validates input with optional item', function (done) {
+
+            var schema = Joi.array().ordered([Joi.string().required(), Joi.number().required(), Joi.number()]);
+            var input = ['s1', 2, 3];
+
+            schema.validate(input, function (err, value) {
+
+                expect(err).to.not.exist();
+                expect(value).to.deep.equal(['s1', 2, 3]);
+                done();
+            });
+        });
+
+        it('validates input without optional item', function (done) {
+
+            var schema = Joi.array().ordered([Joi.string().required(), Joi.number().required(), Joi.number()]).sparse(true);
+            var input = ['s1', 2, undefined];
+
+            schema.validate(input, function (err, value) {
+
+                expect(err).to.not.exist();
+                expect(value).to.deep.equal(['s1', 2, undefined]);
+                done();
+            });
+        });
+
+        it('validates input without optional item in a sparse array', function (done) {
+
+            var schema = Joi.array().ordered([Joi.string().required(), Joi.number(), Joi.number().required()]).sparse(true);
+            var input = ['s1', undefined, 3];
+
+            schema.validate(input, function (err, value) {
+
+                expect(err).to.not.exist();
+                expect(value).to.deep.equal(['s1', undefined, 3]);
+                done();
+            });
+        });
+
+        it('validates when input matches ordered items and matches regular items', function (done) {
+
+            var schema = Joi.array().ordered([Joi.string().required(), Joi.number().required()]).items(Joi.number());
+            var input = ['s1', 2, 3, 4, 5];
+            schema.validate(input, function (err, value) {
+
+                expect(err).to.not.exist();
+                expect(value).to.deep.equal(['s1', 2, 3, 4, 5]);
+                done();
+            });
+        });
+
+        it('errors when input does not match ordered items', function (done) {
+
+            var schema = Joi.array().ordered([Joi.number().required(), Joi.string().required()]);
+            var input = ['s1', 2];
+            schema.validate(input, function (err, value) {
+
+                expect(err).to.exist();
+                expect(err.message).to.equal('"value" at position 0 fails because ["0" must be a number]');
+                done();
+            });
+        });
+
+        it('errors when input has more items than ordered items', function (done) {
+
+            var schema = Joi.array().ordered([Joi.number().required(), Joi.string().required()]);
+            var input = [1, 's2', 3];
+            schema.validate(input, function (err, value) {
+
+                expect(err).to.exist();
+                expect(err.message).to.equal('"value" at position 2 fails because "value" is not allowed');
```

Comment:
I think we should create a new error for this one, it's not very explicit what went wrong with this.